staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -784,6 +789,8 @@ func (c *Cacher) dispatchEvents() {
 			// Never send a bookmark event if we did not see an event here, this is fine
 			// because we don't provide any guarantees on sending bookmarks.
 			if lastProcessedResourceVersion == 0 {
+				// pop expired watchers in case there has been no update
+				c.bookmarkWatchers.popExpiredWatchers()
```

Comment:
do we also need to pop expired bookmark watchers on the other handler paths that return (line 783 and 807)?

Reply:
by only removing watchers when there's no resource change, does that mean a constantly changing resource (like endpoints) would still effectively leak?

Reply:
how expensive is popExpiredWatchers to call? is it concerning to call it approximately once a second?

Reply:
> do we also need to pop expired bookmark watchers on the other handler paths that return (line 783 and 807)?

I don't think so - those effectively can happen only when we're closing cacher.

> by only removing expired watchers when there's no resource change, does that mean a constantly changing resource (like endpoints) would still effectively leak?

not sure I understand your question - can you please clarify?

> how expensive is popExpiredWatchers to call? is it concerning to call it approximately once a second?

cheap enough - we're calling it every 1s anyway as part of dispatchEvent call (somewhere down the stack)

Reply:
For the two places Jordan mentioned, c.dispatchEvent would handle expired watchers (via startDispatching).
So there is no source for leak after this fix goes in.

I also thought about whether the expired bookmark watchers need to be popped for every period in case there is no update.
We can use a counter - beyond a threshold, say 5, we pop the watchers.

I am open to suggestion in this regard.

Reply:
>> by only removing expired watchers when there's no resource change, does that mean a constantly changing resource (like endpoints) would still effectively leak?
>
> not sure I understand your question - can you please clarify?

I misread the code... this is skipping pruning expired watchers if this cacher has **never** received a watch event for the resource type. Same question exists, but in the opposite direction now. By only removing expired watchers when there's a resource change, does that mean an almost never-changing resource (like APIService) would still effectively leak?

Reply:
Jordan:
I see you modified previous comment
Do you have any question ?